File: 백준/Silver/1254. 팰린드롬 만들기/팰린드롬 만들기.cc
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> getPi(string p) {
    int m = (int)p.size(), j = 0;
    vector<int> pi(m, 0);
    for (int i = 1; i < m; i++) {
        while (j > 0 && p[i] != p[j])
            j = pi[j - 1];
        if (p[i] == p[j])
            pi[i] = ++j;
    }
    return pi;
}

int main() {
    string s;
    cin >> s;

    string rev_s = s;
    reverse(rev_s.begin(), rev_s.end());

    vector<int> pi = getPi(rev_s);
    int begin = 0, matched = 0, len = s.size();
    while (begin < len) {
        if (matched < len && s[begin + matched] == rev_s[matched]) {
            matched++;
            if (begin + matched == len) {
                break;
            }
        }
        else {
            if (matched == 0) {
                begin++;
            }
            else {
                begin += matched - pi[matched - 1];
                matched = pi[matched - 1];
            }
        }
    }

    cout << len + begin << endl;

    return 0;
}

File: 백준/Silver/1254. 팰린드롬 만들기/팰린드롬 만들기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String str = br.readLine();

        StringBuffer sb = new StringBuffer(str);
        String rev = sb.reverse().toString();

        System.out.println(str.length() + (rev.length() - getMinPellindrom(str, rev)));
    }

    private static int getMinPellindrom(String str, String rev) {
        int[] pi = getPi(rev);
        int n = str.length();
        int m = rev.length();

        int begin = 0;
        int matched = 0;

        while (begin + matched < n) {
            if (matched < m && str.charAt(begin + matched) == rev.charAt(matched)) {
                matched++;
                if (begin + matched == n) {
                    return matched;
                }
            } else if (matched == 0) {
                begin++;
            } else {
                begin += matched - pi[matched - 1];
                matched = pi[matched - 1];
            }
        }

        return 0;
    }

    private static int[] getPi(String rev) {
        int begin = 1;
        int matched = 0;

        int[] pi = new int[rev.length()];
        while (begin + matched< rev.length()) {
            if (rev.charAt(matched) == rev.charAt(begin + matched)) {
                matched++;
                pi[begin + matched - 1] = matched;
            } else if (matched == 0) {
                begin++;
            } else {
                begin += matched - pi[matched - 1];
                matched = pi[matched - 1];
            }
        }

        return pi;
    }
}
